packages/core/src/browser/tree/tree-iterator.ts

```diff
@@ -78,20 +152,69 @@ export class TreeNodeIterator extends AbstractTreeNodeIterator {
 
 }
 
-export class BackwardTreeNodeIterator extends AbstractTreeNodeIterator {
+/**
+ * Unlike other tree iterators, this does not visit all the nodes, it stops once it reaches the root node
+ * while traversing up the tree hierarchy in an inverse pre-order fashion. This is the counterpart of the `TopToBottomTreeIterator`.
+ */
+export class BottomToTopTreeIterator extends AbstractTreeIterator {
 
-    protected doNext(node: ITreeNode): ITreeNode | undefined {
-        const prevSibling = ITreeNode.getPrevSibling(node);
-        const lastChild = this.findLastChild(prevSibling);
+    constructor(protected readonly root: TreeNode, protected readonly options: TreeIterator.Options = TreeIterator.DEFAULT_OPTIONS) {
+        super(root, options);
+    }
+
+    protected iterator(root: TreeNode): IterableIterator<TreeNode> {
+        const doNext = this.doNext.bind(this);
+        return (function* (): IterableIterator<TreeNode> {
+            let next = root;
+            while (next) {
+                yield next;
+                next = doNext(next);
+            }
+        }).bind(this)();
+    }
+
+    protected doNext(node: TreeNode): TreeNode | undefined {
+        const previousSibling = TreeNode.getPrevSibling(node);
+        const lastChild = this.lastChild(previousSibling);
         return lastChild || node.parent;
     }
 
-    protected findLastChild(node: ITreeNode | undefined): ITreeNode | undefined {
-        if (!this.hasChildren(node)) {
+    protected lastChild(node: TreeNode | undefined): TreeNode | undefined {
+        const children = node ? this.children(node) : [];
+        if (this.isEmpty(children)) {
             return node;
         }
-        const lastChild = ICompositeTreeNode.getLastChild(node);
-        return this.findLastChild(lastChild);
+        if (CompositeTreeNode.is(node)) {
+            const lastChild = CompositeTreeNode.getLastChild(node);
+            return this.lastChild(lastChild);
+        }
+        return undefined;
+    }
+
+}
+
+/**
+ * Generator for depth first, pre-order tree traversal iteration.
+ */
+function* preOrder<T>(root: T, children: (node: T) => T[] | undefined, include: (node: T) => boolean = () => true): IterableIterator<T> {
```

Comment:
Could you put global functions under namespaces please? `preOrder` can go under `PreOrderTreeIterator` for example or be a method in `PreOrderTreeIterator`

Reply:
`include` is not used and seems to be redundant since `children` function is capable to do necessary filtering

Reply:
> global functions under namespaces

It is not global, isn't it? It is not exported.

Reply:
at runtime not, but at the dev time, I think, ts will add it to workspace symbols

regardless of it, it could be handy to call for other implementations of the tree iterator

Reply:
Makes sense.

Reply:
> include is not used and seems to be redundant since children function is capable to do necessary filtering

It is a generic generator, I leave as is. (I have applied a few changes based on your other comments on the iterators.)